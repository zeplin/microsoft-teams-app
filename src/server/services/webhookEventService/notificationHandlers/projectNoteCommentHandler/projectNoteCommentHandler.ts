import {
    ProjectNoteCommentCreatedEvent,
    WebhookEvent
} from "@zeplin/sdk";

import { GroupingKeyParams, NotificationHandler } from "../NotificationHandler";
import { MessageCard, commonTeamsCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";

class ProjectNoteCommentHandler extends NotificationHandler<ProjectNoteCommentCreatedEvent> {
    delay = 0;

    private getText(event: ProjectNoteCommentCreatedEvent): string {
        const {
            context: {
                project: {
                    name: projectName
                },
                screen: {
                    name: screenName,
                    variant: screenVariant
                },
                note: {
                    order
                }
            },
            actor: {
                user: {
                    username
                }
            }
        } = event;
        return (screenVariant)
            ? md`**${username as string}** replied to note _#${order}_ on _${screenVariant.value}_ variant of _${screenVariant.group.name}_ in _${projectName}_. ${getRandomEmoji()}`
            : md`**${username as string}** replied to note _#${order}_ on _${screenName}_ screen in _${projectName}_. ${getRandomEmoji()}`;
    }

    private getSectionText(event: ProjectNoteCommentCreatedEvent): string {
        const {
            resource: {
                data: {
                    content: commentContent
                }
            }
        } = event;
        return commentContent;
    }

    private getWebappURL(event: ProjectNoteCommentCreatedEvent): string {
        const {
            context: {
                screen: {
                    id: screenId
                },
                project: {
                    id: projectId
                },
                note: {
                    id: noteId
                }
            },
            resource: {
                id: commentId
            }
        } = event;
        const pathname = `project/${projectId}/screen/${screenId}`;
        const searchParams = {
            did: noteId,
            cmid: commentId
        };
        const utmParams = { utmCampaign: "comment_reply" };

        return getWebAppURL(pathname, searchParams, utmParams);
    }

    private getZeplinAppURI(event: ProjectNoteCommentCreatedEvent): string {
        const {
            context: {
                project: {
                    id: projectId
                },
                screen: {
                    id: screenId
                },
                note: {
                    id: noteId
                }
            },
            resource: {
                id: commentId
            }
        } = event;
        const searchParams = {
            pid: projectId,
            sid: screenId,
            did: noteId,
            cmids: commentId
        };
        const utmParams = { utmCampaign: "comment_reply" };

        return getRedirectURLForZeplinApp("dot", searchParams, utmParams);
    }

    shouldHandleEvent(event: WebhookEvent): event is ProjectNoteCommentCreatedEvent {
        return event.action === "created";
    }

    getGroupingKey({ webhookId, deliveryId }: GroupingKeyParams<ProjectNoteCommentCreatedEvent>): string {
        return `${webhookId}:${deliveryId}`;
    }

    getTeamsMessage(events: ProjectNoteCommentCreatedEvent[]): MessageCard {
        const [event] = events;
        return commonTeamsCard({
            text: this.getText(event),
            section: {
                text: this.getSectionText(event)
            },
            links: [{
                title: "Open in App",
                url: this.getZeplinAppURI(event)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export const projectNoteCommentHandler = new ProjectNoteCommentHandler();
