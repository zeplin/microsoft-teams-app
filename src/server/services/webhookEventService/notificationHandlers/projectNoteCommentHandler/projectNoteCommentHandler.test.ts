import { ProjectNoteCommentCreatedEvent } from "@zeplin/sdk";

import { projectNoteCommentHandler } from "./projectNoteCommentHandler";

type GetDummyEventParams = {
    projectId?: string;
    projectName?: string;
    screenId?: string;
    screenName?: string;
    username?: string;
    noteId?: string;
    noteOrder?: number;
    commentId?: string;
    commentContent?: string;
}

function getDummyEvent({
    projectId = "projectId",
    projectName = "Project Microsoft Teams Integration",
    screenId = "screenId",
    screenName = "Manage Zeplin Connector",
    username = "sertac",
    noteId = "noteId",
    noteOrder = 1,
    commentId = "commentId",
    commentContent = "cok guzel keyfimiz var"
}: GetDummyEventParams = {}): ProjectNoteCommentCreatedEvent {
    return {
        action: "created",
        context: {
            screen: {
                id: screenId,
                name: screenName
            },
            project: {
                id: projectId,
                name: projectName
            },
            note: {
                id: noteId,
                order: noteOrder
            }
        },
        actor: {
            user: {
                username
            }
        },
        resource: {
            id: commentId,
            data: {
                id: commentId,
                content: commentContent
            }
        }
    } as ProjectNoteCommentCreatedEvent;
}

describe("projectNoteHandler", () => {
    describe("getTeamsMessage", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        it("should match snapshot", () => {
            expect(projectNoteCommentHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });
    });
});
