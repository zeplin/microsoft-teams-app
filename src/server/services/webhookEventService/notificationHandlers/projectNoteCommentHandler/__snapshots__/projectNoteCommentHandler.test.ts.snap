// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectNoteHandler getTeamsMessage should match snapshot 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**sertac** replied to note _#1_ on _Manage Zeplin Connector_ screen in _Project Microsoft Teams Integration_. ðŸ—ž",
  "sections": [
    {
      "text": "cok guzel keyfimiz var",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/zeplin/app-redirect?uri=zpl%3A%2F%2Fdot%3Fpid%3DprojectId%26sid%3DscreenId%26did%3DnoteId%26cmids%3DcommentId%26ref%3Dmicrosoft-teams%26utm_source%3Dzeplin%26utm_medium%3Dteams_integration%26utm_campaign%3Dcomment_reply"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/screen/screenId?did=noteId&cmid=commentId&ref=microsoft-teams&utm_source=zeplin&utm_medium=teams_integration&utm_campaign=comment_reply"
        }
      ]
    }
  ]
}
`;

exports[`projectNoteHandler getTeamsMessage should match snapshot with screen variant 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**sertac** replied to note _#1_ on _Dark_ variant of _Manage Zeplin Connector_ in _Project Microsoft Teams Integration_. ðŸ—ž",
  "sections": [
    {
      "text": "cok guzel keyfimiz var",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/zeplin/app-redirect?uri=zpl%3A%2F%2Fdot%3Fpid%3DprojectId%26sid%3DscreenId%26did%3DnoteId%26cmids%3DcommentId%26ref%3Dmicrosoft-teams%26utm_source%3Dzeplin%26utm_medium%3Dteams_integration%26utm_campaign%3Dcomment_reply"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/screen/screenId?did=noteId&cmid=commentId&ref=microsoft-teams&utm_source=zeplin&utm_medium=teams_integration&utm_campaign=comment_reply"
        }
      ]
    }
  ]
}
`;
